docs/architecture/003-handler-implementation.md

```diff
@@ -0,0 +1,552 @@
+# ADR 003: Handler implementation
+
+## Changelog
+* 2020-08-06: Initial proposal
+
+## Context
+
+> This section contains all the context one needs to understand the current state, and why there is a problem. It should be as succinct as possible and introduce the high level idea behind the solution.
+
+TODO
+
+## Decision
+
+In this chapter, we provide recommendation for implementing IBC handlers within the `ibc-rs` crate.
+Concepts are introduced in the order given by a topological sort of their dependencies on each other.
+
+### Events
+
+IBC handlers must be able to emit events which will then be broadcasted via the node's pub/sub mechanism,
+and eventually picked up by the IBC relayer.
+
+A generic interface for events is provided below, where an event is represented
+as a pair of an event type and a list of attributes. An attribute is simply a pair
+of a key and a value, both represented as strings. TODO: Describe event type.
```

Comment:
Instead of this TODO here, maybe we can link to here for an overview (though they are not precisely the same)?

https://github.com/informalsystems/ibc-rs/blob/master/modules/src/events.rs#L15